dir/fs.go

```diff
@@ -0,0 +1,159 @@
+package dir
+
+import (
+	"errors"
+	"io/fs"
+	"os"
+	"path"
+	"path/filepath"
+)
+
+// rootedFS is io.FS implementation used in New.
+// root is the root of the file system tree passed to os.DirFS.
+type rootedFS struct {
+	fs.FS
+	Root string
+}
+
+// NewRootedFS create a rootedFS
+func NewRootedFS(fsys fs.FS, root string) rootedFS {
+	return rootedFS{fsys, root}
+}
+
+// UnionDirFS is a simple union directory file system interface
+type UnionDirFS interface {
+	fs.ReadDirFS
+	GetPath(elem ...string) (string, error)
+}
+
+// NewUnionDirFS create an unionDirFS on the current OS file system
+// by the named dirs
+func NewUnionDirFS(dirs ...string) unionDirFS {
+	rootedDirs := []rootedFS{}
+	for _, dir := range dirs {
+		rootedDirs = append(rootedDirs, rootedFS{
+			FS:   os.DirFS(dir),
+			Root: dir,
+		})
+	}
+	return unionDirFS{Dirs: rootedDirs}
+}
+
+// NewUnionDirFsByRootedFS create an unionDirFS by rootedFS
+func NewUnionDirFsByRootedFS(rootedFsys ...rootedFS) unionDirFS {
+	return unionDirFS{Dirs: rootedFsys}
+}
```

Comment:
Similarly, we may remove this constructor.

Reply:
I use this method in plugin/manager/manager_test.go to new a memory based vfs.

Reply:
If it is for test only, please make it private.
```suggestion
// newUnionDirFsByRootedFS create an unionDirFS by rootedFS
func newUnionDirFsByRootedFS(rootedFsys ...rootedFS) unionDirFS {
	return unionDirFS{Dirs: rootedFsys}
}
```

Reply:
Redesign NewUnionDirFS() to receive an array of RootedFS. Also, provide a NewRootedFS() to create a rootedFS with self-defined fsys. Because UnionDirFS is based on fs.FS, provide the flexibility to use self-defined fsys is easier to testing and using.